File: 1460/MakeTwoArraysEqualByReversingSubarrays.cpp
/*
https://leetcode.com/problems/make-two-arrays-equal-by-reversing-subarrays/description/?envType=daily-question&envId=2024-08-03
*/

class Solution {
public:

    void performeReverse(vector<int>&arr, int bgn, int end){
        vector<int>tmp;
        for(int i=end; i>=bgn; --i){
            tmp.push_back(arr[i]);
        }
        int idx=0;
        for(int i=bgn; i<=end; ++i){
            arr[i]=tmp[idx++];
        }
    }

    bool canBeEqual(vector<int>& target, vector<int>& arr) {
        int sz=target.size();
        for(int i=0; i<sz; ++i){
            if(target[i]!=arr[i]){
                if(i+1>=sz) return false;
                int j=i, bgn=i;
                bool flag=false;
                while(true){
                    if(j>=sz) break;
                    if(arr[j]==target[i]){
                        flag=true;
                        break;
                    }
                    j++;
                }
                if(!flag) return false;
                performeReverse(arr, bgn, j);
            }
        }
        return arr==target;
    }
};

File: 1460/MakeTwoArraysEqualByReversingSubarrays.java
/*
https://leetcode.com/problems/make-two-arrays-equal-by-reversing-subarrays/description/?envType=daily-question&envId=2024-08-03
*/

class Solution {

    public static void performReverseOfSubarray(int[] arr, int beginIndex, int endIndex){
        int lengthToCover=(endIndex-beginIndex)+1;
        int[] temporaryArray=new int[lengthToCover];
        int start=0;
        for(int i=endIndex; i>=beginIndex; --i){
            temporaryArray[start++]=arr[i];
        }
        start=0;
        for(int i=beginIndex; i<=endIndex; ++i){
            arr[i]=temporaryArray[start++];
        }
    }

    public boolean canBeEqual(int[] target, int[] arr) {
        int lengthOfTarget=target.length;
        for(int i=0; i<lengthOfTarget; ++i){
            if(target[i]!=arr[i]){
                if(i+1>=lengthOfTarget) return false;
                int j=i, beginIndex=i;
                boolean isNumberFoundInRestArray=false;
                while(true){
                    if(j>=lengthOfTarget) break;
                    if(arr[j]==target[i]){
                        isNumberFoundInRestArray=true;
                        break;
                    }
                    j++;
                }
                if(!isNumberFoundInRestArray) return false;
                performReverseOfSubarray(arr, beginIndex, j);
            }
        }
        for(int i=0; i<lengthOfTarget; ++i){
            if(arr[i]!=target[i]) return false;
        }
        return true;
    }
}